File: src/main/python/litcoder/contests/contest-2/ActivitySelection.py
"""Solution to com.ujjwal.solutions.litcoder.contests.contest_2.Activity Selection Contest Problem"""

from typing import List


class com.ujjwal.solutions.litcoder.contests.contest_2.Activity:
    """Wrapper class for activity"""

    start_time: int
    end_time: int

    def __init__(self, start_time: int, end_time: int):
        self.start_time = start_time
        self.end_time = end_time

    def get_time_spent(self):
        return self.end_time - self.start_time

    def __str__(self):
        return f"{self.start_time} {self.end_time}"


def solution(activities: List[com.ujjwal.solutions.litcoder.contests.contest_2.Activity]) -> List[com.ujjwal.solutions.litcoder.contests.contest_2.Activity]:
    """Returns array of activities which contains the maximum number of
    non-overlapping activities that can be performed

    Args:
        activities (List[com.ujjwal.solutions.litcoder.contests.contest_2.Activity]): list of activities

    Returns:
        List[com.ujjwal.solutions.litcoder.contests.contest_2.Activity]:
    """
    # sort the activities in ascending order according to their
    # finishing time
    activities.sort(key=lambda activity: activity.end_time)

    act_order: List[com.ujjwal.solutions.litcoder.contests.contest_2.Activity] = [activities[0]]
    for activity in activities[1:]:
        if activity.start_time >= act_order[-1].end_time:
            act_order.append(activity)

    return act_order


testcases = int(input())
for _ in range(testcases):
    start_times = map(int, input().split(" "))
    end_times = map(int, input().split(" "))

    activities = [com.ujjwal.solutions.litcoder.contests.contest_2.Activity(*times) for times in zip(start_times, end_times)]
    for activity in solution(activities):
        print(activity)

File: src/codes/college/litcoder/contests/contest2/ActivitySelection.java
package codes.college.litcoder.contests.contest2;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Scanner;

class Activity {

    final int startTime;
    final int endTime;

    /**
     * Constructs an Activity object with the given start time and end time.
     *
     * @param startTime the start time of the activity
     * @param endTime   the end time of the activity
     */
    public Activity(int startTime, int endTime) {
        this.startTime = startTime;
        this.endTime = endTime;
    }

    /**
     * Calculates and returns the time spent on the activity.
     *
     * @return the time spent on the activity
     */
    public int getTimeSpent() {
        return endTime - startTime;
    }

    /**
     * Returns a string representation of the Activity object.
     *
     * @return a string representation of the Activity object
     */
    @Override
    public String toString() {
        return startTime + " " + endTime;
    }
}

public class ActivitySelection {

    /**
     * Sorts the activities in ascending order according to their finishing time and
     * returns a list of activities in the order they can be performed.
     *
     * @param activities the list of activities to be sorted
     * @return a list of activities in the order they can be performed
     */
    public static List<Activity> solution(List<Activity> activities) {
        // Sort the activities in ascending order according to their finishing time
        activities.sort(Comparator.comparingInt(a -> a.endTime));

        List<Activity> actOrder = new ArrayList<>();
        actOrder.add(activities.getFirst());

        for (int i = 1; i < activities.size(); i++) {
            Activity activity = activities.get(i);
            if (activity.startTime >= actOrder.getLast().endTime) {
                actOrder.add(activity);
            }
        }

        return actOrder;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int activityCount = scanner.nextInt();
        List<Activity> activities = new ArrayList<>();

        for (int i = 0; i < activityCount; i++) {
            int startTime = scanner.nextInt();
            int endTime = scanner.nextInt();
            activities.add(new Activity(startTime, endTime));
        }

        List<Activity> result = solution(activities);
        for (Activity activity : result) {
            System.out.println(activity);
        }

        scanner.close();
    }
}
